Chatkit/View Models/JoinedRoomsViewModel.swift

```diff
@@ -0,0 +1,171 @@
+import Foundation
+
+/// A view model which provides a collection of all rooms joined by the user.
+///
+/// The collection of all rooms joined by the user is sorted by the view model based on the value
+/// of `createdAt` property of last message on the room when such message is available. When a room
+/// does not contain any messages, the view model uses the value of `createdAt` property of the room
+/// to determine the position of the room in the collection.
+public class JoinedRoomsViewModel {
+    
+    // MARK: - Properties
+    
+    private let provider: JoinedRoomsProvider
+    
+    /// The array of all rooms joined by the user.
+    public private(set) var rooms: [Room]
+    
+    /// The object that is notified when the content of the maintained collection of rooms changed.
+    public weak var delegate: JoinedRoomsViewModelDelegate?
+    
+    // MARK: - Initializers
+    
+    /// Designated initializer for the class.
+    ///
+    /// - Parameters:
+    ///     - provider: The joined rooms provider used as the source of data.
+    public init(provider: JoinedRoomsProvider) {
+        self.rooms = []
+        
+        self.provider = provider
+        self.provider.delegate = self
+        
+        self.reload()
+    }
+    
+    // MARK: - Private methods
+    
+    private func reload() {
+        self.rooms = Array(self.provider.rooms)
+        self.sort()
+    }
+    
+    private func sort() {
+        self.rooms.sort { lhs, rhs -> Bool in
+            if let lhsLastMessage = lhs.lastMessage, let rhsLastMessage = rhs.lastMessage {
+                return lhsLastMessage.createdAt < rhsLastMessage.createdAt
+            }
+            else {
+                return lhs.createdAt < rhs.createdAt
+            }
+        }
+    }
+    
+    private func index(of room: Room) -> Int? {
+        return self.rooms.firstIndex { storedRoom -> Bool in
+            return storedRoom.identifier == room.identifier
+        }
+    }
+    
+}
+
+extension JoinedRoomsViewModel: JoinedRoomsProviderDelegate {
+    
+    public func joinedRoomsProvider(_ joinedRoomsProvider: JoinedRoomsProvider, didJoinRoom room: Room) {
+        // TODO: Optimize if necessary
+        
+        self.rooms.append(room)
+        self.sort()
+        
+        guard let index = self.index(of: room) else {
+            return
+        }
+        
+        self.delegate?.joinedRoomsViewModel(self, didAddRoomAtIndex: index, changeReason: .userJoined)
+    }
+    
+    public func joinedRoomsProvider(_ joinedRoomsProvider: JoinedRoomsProvider, didUpdateRoom room: Room, previousValue: Room) {
+        // TODO: Optimize if necessary
+        
+        guard let previousIndex = self.index(of: previousValue) else {
```

Comment:
It would probably be safer (and less confusing) to find rooms in these lists by comparing their IDs, rather than comparing entire objects and trying to ensure we search using the correct previous/current instance.

Reply:
That is exactly what `index(of:)` method is doing.

Reply:
So it is, sorry!

Reply:
I guess my confusion was the explicit use of the old and new values when we know they have the same identifier.